src/hotspot/share/opto/superword.cpp

```diff
@@ -3615,3 +3030,279 @@ bool SuperWord::same_origin_idx(Node* a, Node* b) const {
+    VTransformNode* iv_vtn = get_vtnode_or_wrap_as_input_scalar(_vloop.iv());
+    BasicType p0_bt = _vloop_analyzer.types().velt_basic_type(p0);
+    // If we have subword type, take that type directly. If p0 is some ConvI2L/F/D,
+    // then the p0_bt can also be L/F/D but we need to produce ints for the input of
+    // the ConvI2L/F/D.
+    BasicType element_bt = is_subword_type(p0_bt) ? p0_bt : T_INT;
+    VTransformNode* populate_index = new (_graph.arena()) VTransformPopulateIndexNode(_graph, pack->size(), element_bt);
+    populate_index->set_req(1, iv_vtn);
+    return populate_index;
+  }
+
+  if (same_input != nullptr) {
+    VTransformNode* same_input_vtn = get_vtnode_or_wrap_as_input_scalar(same_input);
+    if (index == 2 && VectorNode::is_shift(p0)) {
+      // Scalar shift count for vector shift operation: vec2 = shiftV(vec1, scalar_count)
+      // Scalar shift operations masks the shift count, but the vector shift does not, so
+      // create a special ShiftCount node.
+      BasicType element_bt = _vloop_analyzer.types().velt_basic_type(p0);
+      juint mask = (p0->bottom_type() == TypeInt::INT) ? (BitsPerInt - 1) : (BitsPerLong - 1);
+      VTransformNode* shift_count = new (_graph.arena()) VTransformShiftCountNode(_graph, pack->size(), element_bt, mask, p0->Opcode());
+      shift_count->set_req(1, same_input_vtn);
+      return shift_count;
+    } else {
+      // Replicate the scalar same_input to every vector element.
+      const Type* element_type = _vloop_analyzer.types().velt_type(p0);
+      if (index == 2 && VectorNode::is_scalar_rotate(p0) && element_type->isa_long()) {
+        // Scalar rotate has int rotation value, but the scalar rotate expects longs.
+        assert(same_input->bottom_type()->isa_int(), "scalar rotate expects int rotation");
+        VTransformNode* conv = new (_graph.arena()) VTransformConvI2LNode(_graph);
+        conv->set_req(1, same_input_vtn);
+        same_input_vtn = conv;
+      }
+      VTransformNode* replicate = new (_graph.arena()) VTransformReplicateNode(_graph, pack->size(), element_type);
+      replicate->set_req(1, same_input_vtn);
+      return replicate;
+    }
+  }
+
+  // The input is neither a pack not a same_input node. SuperWord::profitable does not allow
+  // any other case. In the future, we could insert a PackNode.
+#ifdef ASSERT
+  tty->print_cr("\nSuperWordVTransformBuilder::get_vtnode_vector_input_at_index: index=%d", index);
+  pack->dump();
+  assert(false, "Pack input was neither a pack nor a same_input node");
+#endif
+  ShouldNotReachHere();
+}
+
+VTransformNode* SuperWordVTransformBuilder::get_vtnode_or_wrap_as_input_scalar(Node* n) {
+  VTransformNode* vtn = get_vtnode_or_null(n);
+  if (vtn != nullptr) { return vtn; }
+
+  assert(!_vloop.in_bb(n), "only nodes outside the loop can be input nodes to the loop");
+  vtn = new (_graph.arena()) VTransformInputScalarNode(_graph, n);
+  set_vtnode(n, vtn);
+  return vtn;
```

Comment:
Note: Since we want the loop-internal nodes to be able to reference all inputs as `vtnodes`, we must pack the inputs that are outside the loop also into special `VTransformInputScalarNode`s.